core/src/main/java/io/grpc/internal/SpiffeUtil2.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2024 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.CertificateParsingException;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Will be merged with SpiffeUtil.
+ * Provides utilities to load, extract, and parse SPIFFE ID according the SPIFFE ID standard.
+ *
+ */
+public final class SpiffeUtil2 {
+
+  private static final Logger log = Logger.getLogger(SpiffeUtil2.class.getName());
+
+  private static final Integer URI_SAN_TYPE = 6;
+  private static final String USE_PARAMETER_VALUE = "x509-svid";
+
+  /**
+   * Parses a leaf certificate from the chain to extract unique SPIFFE ID. In case of success,
```

Comment:
Suggested edit: "Returns the SPIFFE ID from the leaf certificate, if present."

Reply:
I'm on the fence here - a reader already sees that the method returns `Optional<SpiffeId>`, so my intent here is to provide a little bit more details about what we;re returning.
WDYT?

Reply:
Changed as discussed offline